src/common/wayfinder/verification/strategies/composite-verifier.ts

```
@@ -0,0 +1,140 @@
+ * limitations under the License.
+ */
+import { PassThrough, Readable } from 'node:stream';
+
+import { DataVerificationStrategy } from '../../../../types/wayfinder.js';
+
+/**
+ * A verification strategy that composes multiple verification strategies together.
+ * This allows for more comprehensive verification by applying multiple verification techniques.
+ * All verifications are run in parallel, and the process stops on the first failure.
+ */
+export class CompositeVerificationStrategy implements DataVerificationStrategy {
+  private readonly verificationStrategies: DataVerificationStrategy[];
+
+  /**
+   * Creates a new CompositeVerificationStrategy
+   *
+   * @param verificationStrategies - An array of verification strategies to apply
+   */
+  constructor({
+    verificationStrategies,
+  }: {
+    verificationStrategies: DataVerificationStrategy[];
+  }) {
+    if (
+      verificationStrategies === undefined ||
+      verificationStrategies.length === 0
+    ) {
+      throw new Error('At least one verification strategy must be provided');
+    }
+    this.verificationStrategies = verificationStrategies;
+  }
+
+  /**
+   * Verifies data using all configured verification strategies in parallel.
+   * If any verification fails, the process stops and the error is propagated.
+   *
+   * @param data - The data to verify (Buffer, Readable, or ReadableStream)
+   * @param txId - The transaction ID for the data
+   * @returns A promise that resolves if all verifications pass, or rejects with the first verification error
+   */
+  async verifyData({
+    data,
+    txId,
+  }: {
+    data: Buffer | Readable | ReadableStream;
+    txId: string;
+  }): Promise<void> {
+    // for Buffer data, reuse the same buffer for all verifications
+    if (Buffer.isBuffer(data)) {
+      await Promise.all(
+        this.verificationStrategies.map(async (strategy, index) => {
+          try {
+            await strategy.verifyData({ data, txId });
```

In thinking about these running in parallel now, and the possibility that one of the verifiers may fail fast, I'm somewhat desiring to see abort controllers passed through to verifyData and shared among implementations...